### PR TITLE
Unique username

### LINKED ISSUE #12: Enforce username to be unique
The username as a part of public user url should be unique when provided on registration.

### PR DESCRIPTION
**Pull request by Washubrnfg85**
Unique username provided by conditional check in AuthController

### CODE REVIEW

File: src/main/resources/templates/login_page.html [hunk 1]
```diff
@@ -22,10 +22,9 @@
 <body>
 <main>
     <div class="mb-4 text-center">
-        <div class="text-success" th:if="${param.logout}">You have been logged out.</div>
-    </div>
-    <div class="mb-4 text-center">
-        <div class="text-warning" th:if="${param.userExists}">This email already in use. Please login.</div>
+        <div class="text-success" th:if="${param.logout}">You have been LOGGED OUT.</div>
+        <div class="text-warning" th:if="${param.emailExists}">This EMAIL already in use. Please login.</div>
+        <div class="text-warning" th:if="${param.usernameExists}">This USERNAME already in use. Please login.</div>
```
Comment (eugenesmkn) [inline src/main/resources/templates/login_page.html]: If username is already in use, then this doesn't mean that the user can login

File: src/main/java/org/linksharing/server/http/ApiRestController.java [hunk 2]
```diff
@@ -11,6 +11,7 @@
 import java.io.FileInputStream;
 import java.io.FileNotFoundException;
 import java.security.Principal;
+import java.util.Optional;
```
Comment (eugenesmkn) [inline src/main/java/org/linksharing/server/http/ApiRestController.java]: Where do you use Optional?

**Review by eugenesimakin — CHANGES_REQUESTED** [review]

File: src/main/java/org/linksharing/server/http/ApiRestController.java (hunk 2 above)
Comment (Washubrnfg85) [reply src/main/java/org/linksharing/server/http/ApiRestController.java]: Crap... Fixed.

File: src/main/resources/templates/login_page.html (hunk 1 above)
Comment (Washubrnfg85) [reply src/main/resources/templates/login_page.html]: Yeah, you're right.
You think it should be text-warning message, but on registration_page? Or it should be invalid-feedback error message on registration_page like the others?

Comment (eugenesmkn) [reply src/main/resources/templates/login_page.html]: Yes, registration page.
You can choose either one.

File: src/main/java/org/linksharing/server/auth/AuthController.java [hunk 3]
```diff
@@ -40,6 +41,10 @@ public String showRegistrationPage(User user) {
     @PostMapping("/register")
     public String registerUser(@Valid User user, BindingResult result) {
 
+        if (userRepository.existsByUsername(user.getUsername())) {
+            result.addError(new FieldError(result.getObjectName(),"username", "Username already in use."));
```
Comment (Washubrnfg85) [inline src/main/java/org/linksharing/server/auth/AuthController.java:47]: Kinda proud of this solution ))

File: src/main/java/org/linksharing/server/auth/AuthController.java [hunk 4]
```diff
@@ -40,6 +41,10 @@ public String showRegistrationPage(User user) {
     @PostMapping("/register")
     public String registerUser(@Valid User user, BindingResult result) {
```
Comment (eugenesmkn) [inline src/main/java/org/linksharing/server/auth/AuthController.java:42]: The logic in this method doesn't look right...

File: src/main/java/org/linksharing/server/http/ApiRestController.java [hunk 5]
```diff
@@ -67,7 +67,14 @@ ResponseEntity<UserProfileDetails> updateProfileDetails(Principal user, @Request
 
     @GetMapping("/profile/picture")
     ResponseEntity<?> getProfilePicture(Principal user) throws FileNotFoundException {
-        File img = new File("../image.jpg");
+
+        String imageUrl = "../image.jpg";
```
Comment (eugenesmkn) [inline src/main/java/org/linksharing/server/http/ApiRestController.java]: Why did you change this method?
Do you want to implement the next task in this PR?

**Review by eugenesimakin — CHANGES_REQUESTED** [review]

File: src/main/java/org/linksharing/server/http/ApiRestController.java (hunk 5 above)
Comment (Washubrnfg85) [reply src/main/java/org/linksharing/server/http/ApiRestController.java]: No, no, no.
Just a little messed in my local branches, and it was too late, I lost my focus et.
Dicard it please.

I'll implement it in corresponding task.

File: src/main/java/org/linksharing/server/auth/AuthController.java (hunk 4 above)
Comment (eugenesmkn) [reply src/main/java/org/linksharing/server/auth/AuthController.java:42]: For example, you call `existsByUsername` twice.

Comment (Washubrnfg85) [reply src/main/java/org/linksharing/server/auth/AuthController.java:42]: Not sure that got you right, but do admit - method was raw. Fixed it. Please review.

**Review by eugenesimakin — APPROVED** [review]
